Provide initial infrastructure for creating PDF ref tests.

sammygill opened the pull request:
#### c8fad15f783eb5b7f84b415d0c2a8c97b24fc786
<pre>
Provide initial infrastructure for creating PDF ref tests.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270044">https://bugs.webkit.org/show_bug.cgi?id=270044</a>
<a href="https://rdar.apple.com/problem/123573535">rdar://problem/123573535</a>

Reviewed by Tim Horton.

This patch provides the very initial infrastructure that we will build
upon for creating PDF ref tests. The main focus was to create a way
for us to test PDF annotations, but we can build upon this to test other
parts of PDF functionality. I added three very basic tests to cover PDF
annotation interaction: selecting radio buttons, checkboxes, and select
menus.

Two APIs were added to window.internals to make this possible:
1.) registerPDFTestCallback
2.) pdfAnnotationRects

registerPDFTestCallback takes two arguments: the HTML element that
corresponds to the plugin and a callback that will get executed once
the PDF document is installed in the plugin. The callback is supposed to
act as the test that we would like to run and should contain all of the
logic involved with testing a particular part of PDF functionality.
This callback is taken and plumbed into the PDFPlugin so that it can
execute once the document is ready so that it can use any other APIs
that return information about the document. This is done by calling into
the corresponding private method on PluginViewBase that is available to
Internals since it is a friend. I chose this approach because I did not
want this to become a public API on PluginViewBase that can be used
arbitrarily. Also, if the document is already loaded in the plugin then
we will execute the callback immediately in order to avoid holding onto
it.

pdfAnotationRects is a simple API that just returns a list
of the geometry of all the annotations in the document in plugin space.
Returning the geometry of all the annotations was the simplest and was
sufficient enough for the tests that were added in this patch. However,
we can build upon this to filter out which annotations should be
retruned based upon what we find useful for testing.

The current approach to creating a PDF ref test is to load a document
in either a iframe or embed and then register a callback that will
contain all of the logic for the test once the document has been loaded
in the plugin. For example, the tests for the annotations in this patch
use UIHelper pdfAnnotationRects to interact with the document.

One current limitation with the initial infrastructure is that you need
to have a reference document with the modified PDF that the test will be
compared against. The annotation tests mutate the document in the test
itself, but the result is compared with a PDF that is mutated to be the
expectation of the test result. I tried to make it so that the reference
could use the legacy plugin, but my initial attempts were fruitless so
opted for the simpler approach for the sake of getting the groundwork
laid out so that we could iterate on it for future patches. Another
alternative improvement to consider besides getting the legacy plugin
to work is to add some additional APIs that gather data on the PDF
document and then have the tests assert against the data with what they
expect it to be.

* LayoutTests/TestExpectations:
* LayoutTests/pdf/unifiedpdf/annotations/checkbox-set-active-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/checkbox-set-active.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/dropdown-select-second-option-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/dropdown-select-second-option.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/radio-buttons-select-second-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/radio-buttons-select-second.html: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-checkbox-active.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-checkbox-inactive.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-dropdown-first-option-selected.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-dropdown-second-option-selected.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-two-radio-buttons-first-active.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-two-radio-buttons-second-active.pdf: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::pdfAnnotationRectsForTesting const):
(WebCore::PluginViewBase::registerPDFTestCallback):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::pdfAnnotationRectsForTesting const):
(WebCore::Internals::registerPDFTest):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::annotationRectsForTesting const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::registerPDFTest):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::annotationRectsForTesting const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::pdfAnnotationRectsForTesting const):
(WebKit::PluginView::registerPDFTestCallback):
* Source/WebKit/WebProcess/Plugins/PluginView.h:

Canonical link: <a href="https://commits.webkit.org/276228@main">https://commits.webkit.org/276228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f60bcb526058646a814b704b52f118a3a1c3a2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43969 "Failed to checkout and rebase branch from PR 25066") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46619 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40042 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46273 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26943 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20431 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20063 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37873 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17288 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17566 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38960 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2030 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40148 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39252 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48196 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18975 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15549 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43110 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38143 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41841 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9802 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20569 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19987 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -435,6 +438,8 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
     "    margin-top: 16px;"
     "}"_s;
+
+    RefPtr<WebCore::VoidCallback> m_pdfTestCallback;
```
Maybe this can go up before the bizarre giant block of CSS?

hortont424 commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -417,6 +417,16 @@ enum AV1ConfigurationRange {
+[
+    ExportMacro=WEBCORE_TESTSUPPORT_EXPORT,
+    JSGenerateToJSObject,
+] dictionary PDFAnnotationRect {
```
Surprised we don't already have a recty type to use here

cdumez replied:
`DOMRect` comes to mind but it has a bit more information than you need :/

sammygill committed:
Provide initial infrastructure for creating PDF ref tests.
https://bugs.webkit.org/show_bug.cgi?id=270044
rdar://problem/123573535

Reviewed by Tim Horton.

This patch provides the very initial infrastructure that we will build
upon for creating PDF ref tests. The main focus was to create a way
for us to test PDF annotations, but we can build upon this to test other
parts of PDF functionality. I added three very basic tests to cover PDF
annotation interaction: selecting radio buttons, checkboxes, and select
menus.

Two APIs were added to window.internals to make this possible:
1.) registerPDFTestCallback
2.) pdfAnnotationRects

registerPDFTestCallback takes two arguments: the HTML element that
corresponds to the plugin and a callback that will get executed once
the PDF document is installed in the plugin. The callback is supposed to
act as the test that we would like to run and should contain all of the
logic involved with testing a particular part of PDF functionality.
This callback is taken and plumbed into the PDFPlugin so that it can
execute once the document is ready so that it can use any other APIs
that return information about the document. This is done by calling into
the corresponding private method on PluginViewBase that is available to
Internals since it is a friend. I chose this approach because I did not
want this to become a public API on PluginViewBase that can be used
arbitrarily. Also, if the document is already loaded in the plugin then
we will execute the callback immediately in order to avoid holding onto
it.

pdfAnotationRects is a simple API that just returns a list
of the geometry of all the annotations in the document in plugin space.
Returning the geometry of all the annotations was the simplest and was
sufficient enough for the tests that were added in this patch. However,
we can build upon this to filter out which annotations should be
retruned based upon what we find useful for testing.

The current approach to creating a PDF ref test is to load a document
in either a iframe or embed and then register a callback that will
contain all of the logic for the test once the document has been loaded
in the plugin. For example, the tests for the annotations in this patch
use UIHelper pdfAnnotationRects to interact with the document.

One current limitation with the initial infrastructure is that you need
to have a reference document with the modified PDF that the test will be
compared against. The annotation tests mutate the document in the test
itself, but the result is compared with a PDF that is mutated to be the
expectation of the test result. I tried to make it so that the reference
could use the legacy plugin, but my initial attempts were fruitless so
opted for the simpler approach for the sake of getting the groundwork
laid out so that we could iterate on it for future patches. Another
alternative improvement to consider besides getting the legacy plugin
to work is to add some additional APIs that gather data on the PDF
document and then have the tests assert against the data with what they
expect it to be.

* LayoutTests/TestExpectations:
* LayoutTests/pdf/unifiedpdf/annotations/checkbox-set-active-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/checkbox-set-active.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/dropdown-select-second-option-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/dropdown-select-second-option.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/radio-buttons-select-second-expected.html: Added.
* LayoutTests/pdf/unifiedpdf/annotations/radio-buttons-select-second.html: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-checkbox-active.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-checkbox-inactive.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-dropdown-first-option-selected.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-dropdown-second-option-selected.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-two-radio-buttons-first-active.pdf: Added.
* LayoutTests/pdf/unifiedpdf/resources/annotation-two-radio-buttons-second-active.pdf: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::pdfAnnotationRectsForTesting const):
(WebCore::PluginViewBase::registerPDFTestCallback):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::pdfAnnotationRectsForTesting const):
(WebCore::Internals::registerPDFTest):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::annotationRectsForTesting const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::registerPDFTest):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::annotationRectsForTesting const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::pdfAnnotationRectsForTesting const):
(WebKit::PluginView::registerPDFTestCallback):
* Source/WebKit/WebProcess/Plugins/PluginView.h:

Canonical link: https://commits.webkit.org/276228@main
